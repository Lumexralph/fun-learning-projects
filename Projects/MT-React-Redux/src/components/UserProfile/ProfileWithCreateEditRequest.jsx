import React from 'react';
import PropTypes from 'prop-types';

import requestHOC from '../HOC/withUserPage';

export const RequestForm = ({
  onChange, onSubmit, message, pathname, request,
}) => (
  <div id="request-modal" className="modal">
    <div className="modal-content">
      <div className="modal-header">
        <h2>FixZit</h2>
      </div>
      <div className="modal-body">
        <h3>{pathname === '/request/edit' ? 'Edit' : 'Create'} request</h3>
        <div>
          {message ? <p id="formPopup">{message}</p> : ''}
        </div>
        <form
          onSubmit={onSubmit}
        >

          <label htmlFor="title">
            <input
              onChange={e => onChange(e)}
              type="text"
              id="request-title"
              key={request !== null ? request.request_title : ''}
              name="title"
              placeholder="Your request title.."
              defaultValue={request !== null ? request.request_title : ''}
              required
            />
          </label>

          <label htmlFor="dept">
              Department
            <select
              name="department"
              onChange={e => onChange(e)}
              id="dept"
              key={request !== null ? request.department : ''}
              defaultValue={request !== null ? request.department : ''}
            >
              <option value="maintenance">Maintenance</option>
              <option value="Repairs">Repairs</option>
            </select>
          </label>


          <label htmlFor="content">
            <textarea
              onChange={e => onChange(e)}
              id="subject"
              name="content"
              placeholder="Tell us about your request.."
              style={{ height: '200px' }}
              key={request !== null ? request.request_content : ''}
              defaultValue={request !== null ? request.request_content : ''}
              required
            />
          </label>

          <button id="sendRequest" className="btn btn-send" type="submit">{pathname === '/request/edit' ? 'Update' : 'Send'} Request</button>

        </form>
      </div>
    </div>
  </div>
);

RequestForm.defaultProps = {
  request: null,
};

RequestForm.propTypes = {
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  message: PropTypes.string.isRequired,
  pathname: PropTypes.string.isRequired,
  request: PropTypes.any,
};

const ProfileWithRequestForm = requestHOC(RequestForm);

export default ProfileWithRequestForm;
