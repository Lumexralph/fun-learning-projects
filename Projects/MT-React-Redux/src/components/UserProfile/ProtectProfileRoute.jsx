import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import getDataFromStorage from '../../helpers/getDataFromStorage';

const ProtectProfileRoute = ({ component: Component, ...rest }) => {
  const admin = getDataFromStorage('admin');

  return (
    <Route
      {...rest}
      render={(props) => {
        if (admin === 'false') {
          return (
            <Component {...props} />
          );
        }
        return (
          <Redirect
            to={{
              pathname: '/login',
              state: { from: props.location },
            }}
          />
        );
      }
      }
    />);
};

export default ProtectProfileRoute;
