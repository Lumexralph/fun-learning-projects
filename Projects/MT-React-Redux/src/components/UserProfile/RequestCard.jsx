import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const RequestCard = ({ title, department, status, id }) => (
  <div className="flip-card">
    <div className="flip-card-inner">
      <div className="flip-card-front">
        <h1>{title}</h1>
      </div>
      <div className="flip-card-back">
        <h1>{title}</h1>
        <p><i className="fa fa-building" /> {department}</p>
        <p><i className="fa fa-flag-checkered" aria-hidden="true" />  {status}</p>
        <p><Link to={`/requests/${id}`}>Click to view</Link></p>
      </div>
    </div>
  </div>
);

RequestCard.propTypes = {
  title: PropTypes.string.isRequired,
  department: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
};

export default RequestCard;
