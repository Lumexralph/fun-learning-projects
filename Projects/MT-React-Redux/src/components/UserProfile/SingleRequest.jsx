import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import requestHOC from '../HOC/withUserPage';

export const SingleRequest = ({ request }) => (
  <div className="request-container">
    <h2>{request.request_title}</h2>
    <h4><i className="fa fa-file-text-o" aria-hidden="true" />  {request.request_content}</h4>
    <p><i className="fa fa-building" /> {request.department}</p>
    <p><i className="fa fa-flag-checkered" aria-hidden="true" /> {request.status}</p>
    <div>
      <button className="btn btn-nav" type="button">
        <Link to="/request/edit" style={{ textDecoration: 'none' }}><i className="fa fa-edit" /></Link>
      </button>
    </div>
  </div>
);

SingleRequest.propTypes = {
  request: PropTypes.object.isRequired,
};

const SingleRequestPage = requestHOC(SingleRequest);

export default SingleRequestPage;
