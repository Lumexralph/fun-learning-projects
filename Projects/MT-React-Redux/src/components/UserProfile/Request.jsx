import React from 'react';
import PropTypes from 'prop-types';

import RequestCard from './RequestCard';

const Request = ({ data }) => {
  if (data.length > 0) {
    const requestCards = data.map(request => (
      <div key={request.request_id}>
        <RequestCard
          title={request.request_title}
          department={request.department}
          status={request.status}
          id={request.request_id}
        />
      </div>
    ));

    return requestCards;
  }

  return (
    <div className="request-card">
      <p>No requests created yet, you can make a request!</p>
    </div>
  );
};

Request.propTypes = {
  data: PropTypes.array.isRequired,
};

export default Request;
