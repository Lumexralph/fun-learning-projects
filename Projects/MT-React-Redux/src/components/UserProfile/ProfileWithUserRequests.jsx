import React from 'react';
import PropTypes from 'prop-types';

import requestHOC from '../HOC/withUserPage';

import Request from './Request';

export const UserRequests = ({ requests }) => (
  <div className="card-container">
    <Request data={requests} />
  </div>
);

UserRequests.propTypes = {
  requests: PropTypes.array.isRequired,
};

export default requestHOC(UserRequests);
