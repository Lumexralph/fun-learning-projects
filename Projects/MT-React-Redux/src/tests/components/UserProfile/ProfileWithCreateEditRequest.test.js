import React from 'react';
import renderer from 'react-test-renderer';

import { RequestForm } from '../../../components/UserProfile/ProfileWithCreateEditRequest';


describe('Test for RequestForm', () => {
  let component = renderer.create(
    <RequestForm
      request={{}}
      onChange={value => value}
      onSubmit={e => e}
      message="This is RequestForm"
      pathname="/request/edit"
    />,
  );
  let tree = component.toJSON();

  it('Should render component', () => {
    expect(tree).toMatchSnapshot();
  });


  it('Should call onChange event on title input field', (done) => {
    // manually trigger the callback
    tree.children[0].children[1].children[2].children[0].children[0].props.onChange();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    done();
  });

  it('Should call onChange event on select options', (done) => {
    // manually trigger the callback
    tree.children[0].children[1].children[2].children[1].children[1].props.onChange();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    done();
  });

  it('Should call onChange event on textarea', (done) => {
    component = renderer.create(
      <RequestForm
        request={null}
        onChange={value => value}
        onSubmit={e => e}
        message=""
        pathname=""
      />,
    );
    // manually trigger the callback
    tree.children[0].children[1].children[2].children[2].children[0].props.onChange();
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    done();
  });
});
