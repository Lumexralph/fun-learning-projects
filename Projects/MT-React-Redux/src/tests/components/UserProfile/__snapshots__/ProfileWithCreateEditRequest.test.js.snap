// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for RequestForm Should call onChange event on select options 1`] = `
<div
  className="modal"
  id="request-modal"
>
  <div
    className="modal-content"
  >
    <div
      className="modal-header"
    >
      <h2>
        FixZit
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <h3>
        Edit
         request
      </h3>
      <div>
        <p
          id="formPopup"
        >
          This is RequestForm
        </p>
      </div>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="title"
        >
          <input
            id="request-title"
            name="title"
            onChange={[Function]}
            placeholder="Your request title.."
            required={true}
            type="text"
          />
        </label>
        <label
          htmlFor="dept"
        >
          Department
          <select
            id="dept"
            name="department"
            onChange={[Function]}
          >
            <option
              value="maintenance"
            >
              Maintenance
            </option>
            <option
              value="Repairs"
            >
              Repairs
            </option>
          </select>
        </label>
        <label
          htmlFor="content"
        >
          <textarea
            id="subject"
            name="content"
            onChange={[Function]}
            placeholder="Tell us about your request.."
            required={true}
            style={
              Object {
                "height": "200px",
              }
            }
          />
        </label>
        <button
          className="btn btn-send"
          id="sendRequest"
          type="submit"
        >
          Update
           Request
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Test for RequestForm Should call onChange event on textarea 1`] = `
<div
  className="modal"
  id="request-modal"
>
  <div
    className="modal-content"
  >
    <div
      className="modal-header"
    >
      <h2>
        FixZit
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <h3>
        Create
         request
      </h3>
      <div>
        
      </div>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="title"
        >
          <input
            defaultValue=""
            id="request-title"
            name="title"
            onChange={[Function]}
            placeholder="Your request title.."
            required={true}
            type="text"
          />
        </label>
        <label
          htmlFor="dept"
        >
          Department
          <select
            defaultValue=""
            id="dept"
            name="department"
            onChange={[Function]}
          >
            <option
              value="maintenance"
            >
              Maintenance
            </option>
            <option
              value="Repairs"
            >
              Repairs
            </option>
          </select>
        </label>
        <label
          htmlFor="content"
        >
          <textarea
            defaultValue=""
            id="subject"
            name="content"
            onChange={[Function]}
            placeholder="Tell us about your request.."
            required={true}
            style={
              Object {
                "height": "200px",
              }
            }
          />
        </label>
        <button
          className="btn btn-send"
          id="sendRequest"
          type="submit"
        >
          Send
           Request
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Test for RequestForm Should call onChange event on title input field 1`] = `
<div
  className="modal"
  id="request-modal"
>
  <div
    className="modal-content"
  >
    <div
      className="modal-header"
    >
      <h2>
        FixZit
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <h3>
        Edit
         request
      </h3>
      <div>
        <p
          id="formPopup"
        >
          This is RequestForm
        </p>
      </div>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="title"
        >
          <input
            id="request-title"
            name="title"
            onChange={[Function]}
            placeholder="Your request title.."
            required={true}
            type="text"
          />
        </label>
        <label
          htmlFor="dept"
        >
          Department
          <select
            id="dept"
            name="department"
            onChange={[Function]}
          >
            <option
              value="maintenance"
            >
              Maintenance
            </option>
            <option
              value="Repairs"
            >
              Repairs
            </option>
          </select>
        </label>
        <label
          htmlFor="content"
        >
          <textarea
            id="subject"
            name="content"
            onChange={[Function]}
            placeholder="Tell us about your request.."
            required={true}
            style={
              Object {
                "height": "200px",
              }
            }
          />
        </label>
        <button
          className="btn btn-send"
          id="sendRequest"
          type="submit"
        >
          Update
           Request
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Test for RequestForm Should render component 1`] = `
<div
  className="modal"
  id="request-modal"
>
  <div
    className="modal-content"
  >
    <div
      className="modal-header"
    >
      <h2>
        FixZit
      </h2>
    </div>
    <div
      className="modal-body"
    >
      <h3>
        Edit
         request
      </h3>
      <div>
        <p
          id="formPopup"
        >
          This is RequestForm
        </p>
      </div>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="title"
        >
          <input
            id="request-title"
            name="title"
            onChange={[Function]}
            placeholder="Your request title.."
            required={true}
            type="text"
          />
        </label>
        <label
          htmlFor="dept"
        >
          Department
          <select
            id="dept"
            name="department"
            onChange={[Function]}
          >
            <option
              value="maintenance"
            >
              Maintenance
            </option>
            <option
              value="Repairs"
            >
              Repairs
            </option>
          </select>
        </label>
        <label
          htmlFor="content"
        >
          <textarea
            id="subject"
            name="content"
            onChange={[Function]}
            placeholder="Tell us about your request.."
            required={true}
            style={
              Object {
                "height": "200px",
              }
            }
          />
        </label>
        <button
          className="btn btn-send"
          id="sendRequest"
          type="submit"
        >
          Update
           Request
        </button>
      </form>
    </div>
  </div>
</div>
`;
